### PR TITLE
Support Unix Domain Sockets

### PR DESCRIPTION
**Pull request by richardwhiuk**
Andy,

This adds support to CompositeMakeService to support Unix domain sockets.

Tested against a microservice internally.

Signed-off-by: Richard Whitehouse <richard.whitehouse@metaswitch.com>

### CODE REVIEW

File: src/composites.rs [hunk 1]
```diff
@@ -239,6 +239,52 @@ where
...
+                }
+                Poll::Pending => {
+                    return Poll::Pending;
+                }
+            }
+        }
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, target: &'a tokio::net::UnixStream) -> Self::Future {
```
Comment (bossmc) [inline src/composites.rs:270]: `_target`?  Doesn't this warn?

File: src/composites.rs [hunk 2]
```diff
@@ -239,6 +239,52 @@ where
...
+        let mut services = Vec::with_capacity(self.0.len());
+        for (path, service) in &mut self.0 {
+            let path: &'static str = path;
+            services.push(
+                service
+                    .call(None)
```
Comment (bossmc) [inline src/composites.rs:276]: We _could_ pass `Some(target.peer_addr())` for consistency with the `TCP` equivalent - many middlewares just ignore the passed `target` so would still work.  Though any that actually care what the target is will not work with UDS sockets...

Comment (bossmc) [reply src/composites.rs:276]: Obviously the inner `Service` would need to be `Service<&'a tokio::net::uds::SocketAddr>`

File: Cargo.toml [hunk 3]
```diff
@@ -24,6 +24,7 @@ http2 = ["hyper/http2"]
 client = ["hyper/client"]
 tcp = ["hyper/tcp"]
 tls = ["native-tls", "openssl", "hyper-openssl", "hyper-tls"]
+uds = ["tokio", "tokio/net"]
```
Comment (bossmc) [inline Cargo.toml:27]: I _think_ you can just write `uds = ["tokio/net"]` (which implies `tokio`)

**Review by bossmc — APPROVED** [review]

File: Cargo.toml (hunk 3 above)
Comment (richardwhiuk) [reply Cargo.toml:27]: I *think* I've seen that not work previously.

Comment (bossmc) [reply Cargo.toml:27]: Fine, it's not like this hurts, happy to leave as is

File: src/composites.rs (hunk 2 above)
Comment (richardwhiuk) [reply src/composites.rs:276]: We could - in practice, existing middleware takes `Option<std::net::SocketAddr>` and the peer address is unlikely to be interesting anyway.

**Pull request merged**